File: c++/graphs/matrix/dfs.cpp
#include <vector>
#include <string>
#include <unordered_set>
using namespace std;

/**
 * The following is a dfs implementation for the numbe of paths.
 * The space complexity is O(N * M) because that is the potentail size of the call stack.
 * The time complexity is O(4^(N * M) because we are going through each of the N * M elements
 *      with a backtracking approach where there are 4 options to choose from each time.
 *      picture a binary search tree here.
 */

class Solution
{
public:
    int countPaths(vector<vector<int>> &grid)
    {
        unordered_set<string> visited;

        return _helper(0, 0, visited, grid);
    }

private:
    int _helper(int r, int c, unordered_set<string> &visited, vector<vector<int>> &grid)
    {

        string cell = to_string(r) + "-" + to_string(c);
        if (visited.find(cell) != visited.end() || r >= grid.size() || c >= grid[0].size() || r < 0 || c < 0 || grid[r][c] == 1)
        {
            return 0;
        }

        else if (r == grid.size() - 1 && c == grid[0].size() - 1)
        {
            return 1;
        }

        visited.insert(cell);
        int count = 0;

        count += _helper(r + 1, c, visited, grid);
        count += _helper(r - 1, c, visited, grid);
        count += _helper(r, c + 1, visited, grid);
        count += _helper(r, c - 1, visited, grid);

        visited.erase(cell);

        return count;
    }
};

File: Python/Graphs/MatrixGraphs/dfs.py
from typing import List

class Solution:
    """ Count paths implementation. 
        Time complexity: O(4 ^(N * M)). N and M are the dimensions of the matrix respectivly. Consider this a tree with 4 possible routes which are potential the length of the entire matrix.
        Space complexity: O(N * M). This is because we are keeping track of the visited which is potentially the size of the entire matrix."""


    def countPaths(self, grid: List[List[int]]) -> int:
        rows = len(grid)
        cols = len(grid[0])
        visited = dict()


        def _helper(grid, visited, r, c):

            def isOutOfBounds() -> bool:
                if (r < 0 or c < 0
                    or r >= rows or c >= cols
                    or (r, c) in visited
                    or grid[r][c] == 1):
                    return True


            def reachedEnd() -> bool:
                if ( r == rows - 1
                    and c == cols - 1):
                    return 1


            if (isOutOfBounds()):
                return 0
            
            elif (reachedEnd()):
                return 1

            
            visited[(r, c)] = 0

            count = 0
            count += _helper(grid, visited, r + 1, c)
            count += _helper(grid, visited, r - 1, c)
            count += _helper(grid, visited, r, c + 1)
            count += _helper(grid, visited, r, c - 1)

            visited.pop((r, c))

            return count

        

        return _helper(grid, visited, 0, 0)